# scGenoByte: A GenoByte Embedding Transformer with Biological Priors for Cell Type Annotation



**scGenoByte** is a unified deep learning framework designed to enhance single-cell representation learning through biologically informed full-gene modeling. 

------

## ğŸŒŸ Key Innovations

- **Biologically Coherent Tokenization Termed Genobytes**
- **Genome-wise Modeling Framework**
- **Proteomic-Driven Semantic Alignment**
- **Pathway-Guided Regularization** 

------

## ğŸ“‚ Repository Structure

```
scGenoByte/
â”œâ”€â”€ visualization/
	â”œâ”€â”€ genobyte_eval_enrich.py
	â”œâ”€â”€ genobyte_eval_ppi.py
	â””â”€â”€ genobyte_protein_alignment.py  
â”œâ”€â”€ util/
â”œâ”€â”€ GenoByte_construction.py   
â”œâ”€â”€ models_pretrain.py        
â”œâ”€â”€ models_finetune.py         
â”œâ”€â”€ engine_pretrain.py         
â”œâ”€â”€ engine_finetune.py
â”œâ”€â”€ main_finetune.py
â”œâ”€â”€ main_pretrain.py
â”œâ”€â”€ preprocess.py
â”œâ”€â”€ sc_patchemb.py
â”œâ”€â”€ mydataset.py
â”œâ”€â”€ enrich_statistical.py
â””â”€â”€ requirements.txt           
```

------

## ğŸ“Š Data Preparation

The required datasets and biological priors for **scGenoByte** are hosted on Quark Drive. You can download the complete data package from the link below:

> ğŸ”— **Data Download Link:** [https://pan.quark.cn/s/fe21d768b090](https://www.google.com/search?q=https://pan.quark.cn/s/fe21d768b090&authuser=1)

### ğŸ“‚ Directory Structure

Plaintext

```
scGenoByte_Data/
â”œâ”€â”€ down-stream datasets/       # Benchmarking datasets for cell type annotation
â”‚   â”œâ”€â”€ Baron/                  
â”‚   â”œâ”€â”€ Lung/                   
â”‚   â”œâ”€â”€ MacParland/             
â”‚   â”œâ”€â”€ Muraro/                 
â”‚   â”œâ”€â”€ Pan-GI/                 
â”‚   â”œâ”€â”€ Segerstolpe/            
â”‚   â”œâ”€â”€ Xin/                    
â”‚   â””â”€â”€ Zheng68K/               
â”œâ”€â”€ esm_embedding/              # Protein squences embedding (via ESM-2/ESM-1b)
â”‚   â”œâ”€â”€ Homo_sapiens.GRCh38.pep.all.clean.fa
â”‚   â”œâ”€â”€ Homo_sapiens.GRCh38.pep.all.fa
â”‚   â”œâ”€â”€ Homo_sapiens.GRCh38.pep.all.gene_symbol_to_embedding_ESM1b.pt
â”‚   â””â”€â”€ Homo_sapiens.GRCh38.pep.all.gene_symbol_to_protein_ID.json
â”œâ”€â”€ HOMOLOGY/                   # Gene paralogy information
â”‚   â””â”€â”€ HOMOLOGY_FILE.txt       
â”œâ”€â”€ model/                      # Model checkpoints
â”‚   â””â”€â”€ pretrain_model/         
â””â”€â”€ ppi_network/                # Biological interaction priors
    â””â”€â”€ format_h_sapiens.csv    
```

------

### ğŸ“ Dataset Descriptions

| **Category**                    | **Component**          | **Description**                                              |
| ------------------------------- | ---------------------- | ------------------------------------------------------------ |
| **Benchmark**                   | `down-stream datasets` | Contains 8 standardized datasets used for cross-validation.  |
| **Protein Sequences Embedding** | `esm_embedding`        | Pre-computed embeddings from ESM models to support topological semantic alignment. |
| **Homology**                    | `HOMOLOGY`             | Parsed homology data used to calculate the edge weights for GenoByte construction. |
| **PPI**                         | `ppi_network`          | protein-protein interaction scores.                          |
| **Weights**                     | `model`                | Pre-trained scGenoByte parameters.                           |

------

## ğŸš€ Quick Start

### 1. Installation

```
git clone git@github.com:yjs193/scGenoByte.git
cd scGenoByte
pip install -r requirements.txt
```

### 2. GenoByte Construction

```
python GenoByte_construction.py
```

### 3. Training & Annotation

**Pre-training:**

```
CUDA_VISIBLE_DEVICES=0 torchrun --nproc_per_node=1 main_pretrain.py \
--data_path your_data_path.h5ad \
--protein_embed_path your_esm_embedding_path.pt \
--batch_size 128 \
--epoch 200 \
--output_dir ./output
CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc_per_node=2 main_pretrain.py ... # multi gpus
```

**Fine-tuning (e.g., Pan-GI dataset):**

```
CUDA_VISIBLE_DEVICES=0 torchrun --nproc_per_node=1 main_finetune.py \  
--data_path your_data_path.h5ad \ 
--finetune your_model_checkpoint_path.h5ad \
--batch_size 128 \
--epoch 100 \
CUDA_VISIBLE_DEVICES=0,1 torchrun --nproc_per_node=2 main_finetune.py ... # multi gpus
```

## ğŸ“§ Contact

For questions, please contact J. Yao (csyjs@mail.scut.edu.cn)